### DISCUSSION TITLE
Parsing behavior

### DISCUSSION THREAD

**Discussion by yackinn** [mikro-orm/mikro-orm#2881] [Q&A]
Simpliefied setup
```ts
@Get()
  async findAll() {
    const reportSearchResult = await this.reportService.findAndCount([1]);  
    
    return reportSearchResult; // reportSearchResult = { id: 1,  ..., user: User{id: 2, name: "test", ...}}
  }
```
Why is the user instance converted into the number 2 in the response?
The .toJSON method returns the serialized instance with all properties.

That's the response.
```json
{
      "id": 1,
      "price": 4100,
      "company": "Toyota",
      "model": "Corolla",
      "year": 1999,
      "lng": 0,
      "lat": 0,
      "mileage": 100000,
      "user": 2                      // missing serialized user instance
    }
```

**Comment by B4nan** [comment]
How can we tell, without knowing what your repository is actually doing?

> The .toJSON method returns the serialized instance with all properties.

`toJSON` of what? the user entity or the root entity that holds the reference?

You probably do not use `populate` hints, so it is not marked as populated (while it can be still loaded, e.g. via `wrap(e).init()`), hence serialized as just the PK.

**Reply by yackinn** [reply]
Right. I debugged further. The user is fetched at a different place and attached to the request. The report entities that hold a reference to the loaded user contain the entire user object. Other reports only have the users id property.

Does that mean whatever is loaded during the request is stored in memory and available to any orm operation (query)?

I assumed the query result only yields the information that is returned by the issued query.

Is it possible to change the parsed format?

It shouldn't just be the id.

```ts
const report = { id: 1, model: "test", user: 8 }
// should be
const report = { id: 1, model: "test", user: {id : 8} }
```

**Reply by B4nan** [reply]
Parsing != serialization, its actually the very opposite.

> The user is fetched at a different place and attached to the request.

How? You might just do `wrap(user).populated()`, that often helps with this.

> The report entities that hold a reference to the loaded user contain the entire user object. Other reports only have the users id property.

This sounds exactly like the reason why we use populate hints - so we know what you want to have populated and what not, even if you load it. What you are asking for sounds like an unstable API response (one row can have user relation loaded, while rest not), to me this is an antipattern. Having unloaded entities to serialize as objects would make this a tad bit better, still far away from expected.

> It shouldn't just be the id.

You can use property serializer. I am open for a PR that allows this globally, but it needs to be opt-in.

**Reply by yackinn** [reply]
I think my description was confusing. I try to use a constructed example to clarify what I mean.

```ts
@Get()
  async findAll(@Query('fetch') fetchAllUsers: boolean) {
    if (fetchAllUsers) {
      await this.userService.findAll()
    }
    const reportSearchResult = await this.reportService.findAndCount();
    
    return reportSearchResult;
  }
```

The find all users query just issue a query and doesn't do anyhting with the returned data.
Yet the user data are avalaible on the report query.
That's what was confusing.

fetchAllUsers=true
```ts
reportSearchResult = [
	{ id: 1, brand: "test", user: { id:2, name: "Max", age: 22 } },
	{ id: 2, brand: "test", user: { id:1, name: "Joe", age: 30 } }
]
```

fetchAllUsers=false
```ts
reportSearchResult = [
	{ id: 1, brand: "test", user: { id:2 } },
	{ id: 2, brand: "test", user: { id:1 } }
]
```

At what point is the user property converted into a number as the response yields {user: 2}?

Why is Reference<User> {id: 2} serialized into 2?

**Reply by B4nan** [reply]
Yeah this is exactly why we have populate hints, you should use that instead of preloading all users via separate calls. Population is smart and performant, it will fire only one query for each relation, and unlike your version, it will actually populate only those users that are referenced, not necessarily all of them (loading everything is another antipattern, you definitely dont want that unless you are working on some toy project - this is literally a production database killer).

```ts
  async findAll(@Query('fetch') fetchAllUsers: boolean) {
    if (fetchAllUsers) {
      return this.reportService.findAndCount({ populate: ['user'] });
    }

    return this.reportService.findAndCount();
  }
```

**Reply by yackinn** [reply]
Sure. If I don't want to populate the relation at all but keep the foreign key as the id is it correct to override toJSON()?
It skimmed through the repo and found the entity manager and the toObject() method.

```ts

@ManyToOne(() => User, { eager: false })
user: User | null;

toJSON?(strict = false, strip = [], ...args: any[]): { [p: string]: any } {
    // how to prevent the user property from being manipulated by toObject()
	// Is this what you meant?
	wrap(this.user).populated()

    const o = wrap(this, true).toObject(...args)

    return o;
  }
```

target
```ts
const report = {
	id: 1,
	...
	user: {
		id: 2
	}
}
```

not
```ts
const report = {
	id: 1,
	...
	user: 2
}
```

**Reply by B4nan** [reply]
Yes, `toJSON` is an extension point, you can implement custom one, where you can use `toObject()` (or alternatively `toPOJO()` that ignores populate hints) and modify its shape before you return. Both of those methods will give you an object without any circular references.

Or use the property serializer as suggested earlier.

https://mikro-orm.io/docs/serializing#property-serializers

**Reply by yackinn** [reply]
I'm not sure what's the best way to achieve this.

I only want to manipulate the serialization of the entity when it's not populated.
At this point it's only a wrapped entity (just a reference).

```ts
@ManyToOne({ serializer: value => ({id: value.id}) })
  author: Author;
```
This will always change the serialization

Is there a proper way to do this?

**Reply by B4nan** [reply]
Yes, proper way is to add an option to control this globally and adjust the serialization process. I already said PR is welcome.
